File: windows/NeoTUN.Core/Models/VpnProfile.cs
using System.Text.Json.Serialization;

namespace NeoTUN.Core.Models;

public record VpnProfile(
    string Id,
    string Name,
    VpnProtocol Protocol,
    string Server,
    int Port,
    VpnCredentials Credentials,
    VpnSettings Settings,
    long CreatedAt,
    long? LastUsed = null
)
{
    public VpnProfile() : this(
        Guid.NewGuid().ToString(),
        string.Empty,
        VpnProtocol.VMess,
        string.Empty,
        0,
        new VpnCredentials.VMess(string.Empty, 0, "auto"),
        new VpnSettings(),
        DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
    ) { }
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum VpnProtocol
{
    VMess,
    VLess,
    Trojan,
    Shadowsocks
}

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(VMess), "vmess")]
[JsonDerivedType(typeof(VLess), "vless")]
[JsonDerivedType(typeof(Trojan), "trojan")]
[JsonDerivedType(typeof(Shadowsocks), "shadowsocks")]
public abstract record VpnCredentials
{
    public record VMess(
        string UserId,
        int AlterId = 0,
        string Security = "auto"
    ) : VpnCredentials;

    public record VLess(
        string UserId,
        string? Flow = null,
        string Encryption = "none"
    ) : VpnCredentials;

    public record Trojan(
        string Password
    ) : VpnCredentials;

    public record Shadowsocks(
        string Method,
        string Password
    ) : VpnCredentials;
}

public record VpnSettings(
    string Network = "tcp",
    string Security = "none",
    TlsSettings? TlsSettings = null,
    WebSocketSettings? WsSettings = null,
    RealitySettings? RealitySettings = null
);

public record TlsSettings(
    string? ServerName = null,
    bool AllowInsecure = false,
    string[]? Alpn = null
)
{
    public TlsSettings() : this(null, false, Array.Empty<string>()) { }
}

public record WebSocketSettings(
    string Path = "/",
    Dictionary<string, string>? Headers = null
)
{
    public WebSocketSettings() : this("/", new Dictionary<string, string>()) { }
}

public record RealitySettings(
    string PublicKey,
    string ShortId,
    string ServerName,
    string Fingerprint = "chrome"
);

File: android/app/src/main/java/com/neotun/android/models/VpnProfile.kt
package com.neotun.android.models

import kotlinx.serialization.Serializable
import java.io.Serializable as JavaSerializable

@Serializable
data class VpnProfile(
    val id: String = generateId(),
    val name: String,
    val protocol: VpnProtocol,
    val server: String,
    val port: Int,
    val credentials: VpnCredentials,
    val settings: VpnSettings = VpnSettings(),
    val createdAt: Long = System.currentTimeMillis(),
    val lastUsed: Long? = null
) : JavaSerializable

private fun generateId(): String {
    return "profile_${System.currentTimeMillis()}_${(1000..9999).random()}"
}

@Serializable
enum class VpnProtocol : JavaSerializable {
    VMESS,
    VLESS,
    TROJAN,
    SHADOWSOCKS
}

@Serializable
sealed class VpnCredentials : JavaSerializable {
    @Serializable
    data class VMess(
        val userId: String,
        val alterId: Int = 0,
        val security: String = "auto"
    ) : VpnCredentials()
    
    @Serializable
    data class VLess(
        val userId: String,
        val flow: String? = null,
        val encryption: String = "none"
    ) : VpnCredentials()
    
    @Serializable
    data class Trojan(
        val password: String
    ) : VpnCredentials()
    
    @Serializable
    data class Shadowsocks(
        val method: String,
        val password: String
    ) : VpnCredentials()
}

@Serializable
data class VpnSettings(
    val network: String = "tcp",
    val security: String = "none",
    val tlsSettings: TlsSettings? = null,
    val wsSettings: WebSocketSettings? = null,
    val realitySettings: RealitySettings? = null
) : JavaSerializable

@Serializable
data class TlsSettings(
    val serverName: String? = null,
    val allowInsecure: Boolean = false,
    val alpn: List<String> = emptyList()
) : JavaSerializable

@Serializable
data class WebSocketSettings(
    val path: String = "/",
    val headers: Map<String, String> = emptyMap()
) : JavaSerializable

@Serializable
data class RealitySettings(
    val publicKey: String,
    val shortId: String,
    val serverName: String,
    val fingerprint: String = "chrome"
) : JavaSerializable